tests/test_models/test_ddpm.py

```diff
@@ -0,0 +1,612 @@
+from copy import deepcopy
+
+import numpy as np
+import pytest
+import torch
+
+from mmgen.models.builder import build_model
+from mmgen.models.diffusions import (BasicGaussianDiffusion,
+                                     UniformTimeStepSampler)
+from mmgen.models.diffusions.utils import _get_label_batch, _get_noise_batch
+
+
+class TestBasicGaussianDiffusion:
+
+    @classmethod
+    def setup_class(cls):
+        cls.config = dict(
+            type='BasicGaussianDiffusion',
+            num_timesteps=10,
+            betas_cfg=dict(type='cosine'),
+            train_cfg=None,
+            test_cfg=None)
+        cls.denoising = dict(
+            type='DenoisingUnet',
+            image_size=32,
+            in_channels=3,
+            base_channels=128,
+            resblocks_per_downsample=1,
+            attention_res=[16, 8],
+            use_scale_shift_norm=True,
+            dropout=0.3,
+            num_heads=1,
+            use_rescale_timesteps=True,
+            output_cfg=dict(mean='eps', var='learned_range'),
+        )
+        cls.sampler = dict(type='UniformTimeStepSampler')
+        cls.ddpm_loss = [
+            dict(
+                type='DDPMVLBLoss',
+                rescale_mode='constant',
+                rescale_cfg=dict(scale=1),
+                data_info=dict(
+                    mean_pred='mean_pred',
+                    mean_target='mean_posterior',
+                    logvar_pred='logvar_pred',
+                    logvar_target='logvar_posterior'),
+                log_cfgs=[
+                    dict(
+                        type='quartile',
+                        prefix_name='loss_vlb',
+                        total_timesteps=1000),
+                    dict(type='name')
+                ]),
+            dict(
+                type='DDPMMSELoss',
+                log_cfgs=dict(
+                    type='quartile',
+                    prefix_name='loss_mse',
+                    total_timesteps=1000),
+            )
+        ]
+
+    def test_diffusion(self):
+        # test build model
+        cfg = deepcopy(self.config)
+        cfg['denoising'] = self.denoising
+        cfg['timestep_sampler'] = self.sampler
+        cfg['ddpm_loss'] = None
+        diffusion = build_model(cfg)
+        assert isinstance(diffusion, BasicGaussianDiffusion)
+        assert isinstance(diffusion.sampler, UniformTimeStepSampler)
+        assert not diffusion.use_ema
+
+        # test build model --> parse train_cfg with ema
+        cfg = deepcopy(self.config)
+        cfg['denoising'] = self.denoising
+        cfg['timestep_sampler'] = self.sampler
+        cfg['ddpm_loss'] = None
+        cfg['train_cfg'] = dict(use_ema=True)
+        diffusion = build_model(cfg)
+        assert isinstance(diffusion, BasicGaussianDiffusion)
+        assert isinstance(diffusion.sampler, UniformTimeStepSampler)
+        assert diffusion.use_ema is not None
```

Comment:
Why assert `diffusion.use_ema` is not None?

Reply:
Because in line79, we set `use_ema=True` in `train_cfg` for test ema setting.